DietPi-Software | Add Firefox Sync Server

CedArctic opened the pull request:
**Status**: Testing
- [x] Get installation and sync working

**Commit list/description**:
- DietPi-Software | Added Firefox Sync Server by Mozilla
- All other commits are fixes to the initial

**Sources**:
https://mozilla-services.readthedocs.io/en/latest/howtos/run-sync-1.5.html
https://github.com/mozilla-services/syncserver

**Description**:
This is Mozilla's Firefox Sync Server which manages syncing FF instance bookmarks, history, tabs and passwords across devices. Out of the box it runs on a Python server for small loads and can be configured to run behind nginx or Apache.

I've added an installation in dietpi-software to setup the basic python server (no nginx or apache yet, as LAN with only a handful of devices was the target). I'm currently running it on my Pi and managed to sync my devices. Please feel free to take a look and let me know on what can be improved.

Edit: Sync working on mobile devices as well

CedArctic committed:
Added Firefox Sync Server to Optimized Software

CedArctic committed:
Fixes to Firefox Sync configuration and serving

CedArctic committed:
Firefox Sync fixes

- Fixed long description crashing dietpi-software
- Fixed if syntax error on 5320
- Fixed bash variable initialization syntax on 11233
- Enabled writing host IP to config
- Fixed service file
- Enabled service to start on boot

CedArctic committed:
Firefox Sync DB fixes

Fixed ffsync user creation and directory ownership for the database

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1032,6 +1032,17 @@ DietPi-Software will decrypt and use it for software installs. You can change it
+		aSOFTWARE_CATEGORY_INDEX[$software_id]=4
+		aSOFTWARE_ONLINEDOC_URL[$software_id]='p=42#p42'
+		aSOFTWARE_REQUIRES_BUILDESSENTIAL[$software_id]=1
+		aSOFTWARE_REQUIRES_GIT[$software_id]=1
```
Git is actually not required here, the GitHub archive can be downloaded directly, as we do in other cases: https://github.com/mozilla-services/syncserver/archive/master.tar.gz
As long as there is no internal updater that makes use of Git, and one does not plan to develop the code, we can skip it.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5293,6 +5304,34 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+			INSTALL_URL_ADDRESS='https://github.com/mozilla-services/syncserver'
+
+			# If there is an existing installation update it as per Mozilla's instructions, else fresh install
+			if [[ -d /etc/firefox-sync ]]; then
```
Generally /etc is not intended for programs, but for config files only. What do you think about using `/opt/firefox-sync`, or if it contains variable config files that cannot be placed into a different dir, `$G_FP_DIETPI_USERDATA/firefox-sync`?

Generally best DietPi practice would be to have the program dir, that can be fully purged on reinstall in /opt and all config files in $G_FP_DIETPI_USERDATA, but not possible in all cases.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5293,6 +5304,34 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+				git pull
+				git stash pop
+			else
+				git clone "$INSTALL_URL_ADDRESS" $G_FP_DIETPI_USERDATA/firefox-sync
```
$G_FP_DIETPI_USERDATA should not be used as temporary download dir. DietPi scripts navigate into `/tmp/$G_PROGRAM_NAME`, here `/tmp/DietPi-Software`, hence a tmpfs which is perfectly fine as temporary download dir. So it could be:
```
git clone "$INSTALL_URL_ADDRESS"
mv syncserver /etc/firefox-sync
```
... respectively an alternative download method (not Git) and alternative target dir as suggested above.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11213,54 @@ _EOF_
+			# Modify config file
+			ffsync_ip=$(ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | cut -f1 -d'/')
+			dietpiUsrDataEsc=$(sed 's|/|\\/|g' <<< $G_FP_DIETPI_USERDATA)
+			sed -i.bak "s/host = 0.0.0.0/host = $ffsync_ip/" /etc/firefox-sync/syncserver.ini
```
Ah here we have the config file. This cannot be moved to a different location outside the actual program/install dir, can it?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11213,54 @@ _EOF_
+			cat << _EOF_ > /etc/systemd/system/firefox-sync.service
+[Unit]
+Description=Firefox Sync Server (DietPi)
+After=dietpi-boot.service network.target
```
Since Stretch I would use `network-online.target` instead of `network.target`:
```
Wants=network-online.target
After=network-online.target dietpi-boot.service
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11213,54 @@ _EOF_
+			# User
+			local usercmd='useradd -rMU'
+			getent passwd ffsync &> /dev/null && usercmd='usermod -a'
+			$usercmd -G dietpi -d /opt/ffsync -s $(command -v nologin) ffsync
```
Users home dir should be an existing one, I guess? If config file can be moved to outside the program/install dir, I would use this, else the program/install dir. If this will be `/opt/ffsync` (as suggested above), this is fine already, of course :wink:.

MichaIng commented:
@CedArctic 
Many thanks for your work! Very great implementation. I did a few suggestions, most importantly about the install dir, which should not be /etc. I know we use it in some other cases as well, ancient remain that we plan to change by times. Everything else pretty minor.

I'll see if I find time for some testing and merging before v6.29 release, else it will be the very first thing for v6.30!

CedArctic committed:
Firefox Sync General Fixes

- Changed install dir to /opt/firefox-sync
- Removed git updating logic
- Download_Install is used instead of git
- DietPi user data dir is not used as temporary storage during installation
- ffsync user home dir is set to /opt/firefox-sync
- Use database and config file are placed in dietpi_userdata/firefox-sync
- Added logic to modify the makefile to make it use the syncserver.ini config file located in dietpi_userdata/firefox-sync/syncserver.ini

CedArctic commented:
@MichaIng thank you for the swift and helpful response! It's a pleasure contributing to the project. 
I've made changes according to your above feedback and even managed to move the config file out of the installation directory. Would be glad to hear your thoughts and further feedback.

In the mean time if you're testing it you can check out the Quickstart instructions here:
https://github.com/mozilla-services/syncserver
https://mozilla-services.readthedocs.io/en/latest/howtos/run-sync-1.5.html
Just a couple of notes:
- When adding http://localhost:5000/token/1.0/sync/1.5 to Firefox, make sure to not ommit the **token** portion (the default URL doesn't have it and can be deceiving 😛 )
- If sync is not working and the service logs print out something about a URL mismatch, try changing force_wsgi_environ to true in $G_FP_DIETPI_USERDATA/firefox-sync/syncserver.ini

CedArctic commented:
@MichaIng wanted to chime in with some feedback: I've tested FFSync extensively in the past few weeks running on my Pi2 and everything seems to be working as expected across all my devices (Win, Mac, Linux, Android)

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5293,6 +5303,27 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+			
+			# Build
+			cd /opt/firefox-sync
+			sudo make build
```
I tested on x86_64 VM:
- `libffi-dev` and `libssl-dev` were indeed not required to build all Python modules, but `python-dev`, as usual, to build wheels for some, `umemcache` in particular until the build failed. Was
```
DEPS_LIST='python-virtualenv libffi-dev libssl-dev'
```
intended to be
```
DEPS_LIST='python-virtualenv python-dev'
```
?
- `sudo` is not required here as dietpi-software always runs as root 😉.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+			mkdir -p $G_FP_DIETPI_USERDATA/firefox-sync
+
+			# Create database if it doesn't exist
+			if [[ ! -a $G_FP_DIETPI_USERDATA/firefox-sync/FF-Sync-DB.db ]]; then
```
At best use `[[ ! -f ...` here as the database file must be a regular file (or a symlink pointing to a regular file) and not e.g. a directory. The `sqlite3` command below would throw an error then but at least users then have an output to start debugging with 😉.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+
+			# Copy and modify config file
+			cp /opt/firefox-sync/syncserver.ini $G_FP_DIETPI_USERDATA/firefox-sync
+			ffsync_ip=$(ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | cut -f1 -d'/')
```
Use `local ffsync_ip=$(mawk 'NR==4' /run/dietpi/.network)` here, so we have a consistent method do get the IP of the main network adapter. Generally create `local` variables which are only valid for the current function, same below for `dietpiUsrDataEsc` (but this can be skipped as below comment 😉).

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+			sed -i.bak "s/localhost:5000\//$ffsync_ip:5000\//" $G_FP_DIETPI_USERDATA/firefox-sync/syncserver.ini
+			sed -i.bak "s/#sqluri = sqlite:\/\/\/\/tmp\/syncserver.db/sqluri = sqlite:\/\/\/$dietpiUsrDataEsc\/firefox-sync\/FF-Sync-DB.db/" $G_FP_DIETPI_USERDATA/firefox-sync/syncserver.ini
+			# Modify make file to point to external config
+			sed -i.bak "s/--paste .\/syncserver.ini/--paste $dietpiUsrDataEsc\/firefox-sync\/syncserver.ini/"
```
Let's touch the settings file only if it does not exist, so we do not override users custom settings:
```
local ffsync_conf="$G_FP_DIETPI_USERDATA/firefox-sync/syncserver.ini"
if [[ ! -f $ffsync_conf ]]; then

	cp -a /opt/firefox-sync/syncserver.ini $ffsync_conf
...
```
Then to apply settings, we can use `G_CONFIG_INJECT` (which automatically escapes all special characters etc):
```
...
	G_CONFIG_INJECT 'host =' "host = $ffsync_ip" $ffsync_conf
	G_CONFIG_INJECT 'port =' 'port = 5000' $ffsync_conf # To be failsafe if the default changes in future releases
	G_CONFIG_INJECT 'public_url =' "public_url = http://$ffsync_ip:5000/" $ffsync_conf
	G_CONFIG_INJECT 'sqluri = sqlite:' "sqluri = sqlite:///$G_FP_DIETPI_USERDATA/firefox-sync/FF-Sync-DB.db" $ffsync_conf
	sed -i "s|--paste ./syncserver.ini|--paste $ffsync_conf|" /opt/firefox-sync/Makefile # No slash escaping required when using | as delimiter

}
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+[Service]
+User=ffsync
+WorkingDirectory=/opt/firefox-sync
+ExecStart=make serve
```
Use absolute path:
```
ExecStart=$(command -v make) serve
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+_EOF_
+
+		# Permissions
+		chown -R ffsync:dietpi $G_FP_DIETPI_USERDATA/firefox-sync
```
Does ffsync actually require access to any files of "dietpi" group? As long as it only writes to `$G_FP_DIETPI_USERDATA/firefox-sync`, it can use it's own group.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11174,6 +11205,59 @@ _EOF_
+		chown -R ffsync:dietpi $G_FP_DIETPI_USERDATA/firefox-sync
+
+		# Enable startup on boot by default
+		systemctl enable firefox-sync
```
We can add it to DietPi-Services control, so it will be stopped on DietPi updates, backups and software installs, what you think?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -12838,6 +12922,21 @@ _EOF_
+
+			fi
+			getent passwd ffsync &> /dev/null && userdel -rf ffsync
+			[[ -d '/opt/firefox-sync' ]] && rm -R /opt/firefox-sync
```
Shall we remove the config/data dir `$G_FP_DIETPI_USERDATA/firefox-sync` as well?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -5293,6 +5303,27 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+			# Download & Install
+			INSTALL_URL_ADDRESS='https://github.com/mozilla-services/syncserver/archive/master.tar.gz'
+			Download_Install "$INSTALL_URL_ADDRESS"
+			mv syncserver /opt/firefox-sync
```
Since `/opt/firefox-sync` does not contain settings/database anymore, we can remove it for reinstalls, if existing, else `mv` would fail:
```
[[ -d '/opt/firefox-sync' ]] && rm -R /opt/firefox-sync
```

MichaIng commented:
Btw ignore `CodeFactor`. I already fixed things in dev branch, so you could satisfy it by merging current dev branch. However I'll merge regardless of course.

CedArctic committed:
Merge pull request #1 from MichaIng/dev

Merge latest dev changes

CedArctic committed:
Changes for Firefox Sync Server integration

- Fixed dependencies list
- Emptied install dir if it existed before install
- Removed sudo where not needed
- Checked if database existed using -f rather than -a
- Modified config file only if a use modified one does not exist
- Used absolute path for make serve
- Switched ffsync user to its own user group
- Removed the user data directory on uninstall

CedArctic commented:
@MichaIng thanks for the feedback and guidance. I've reviewed your proposed changes and addressed them all except adding FFsync to DietPi-Services control which seems like a great idea, however I am not sure how to go about implementing it.

MichaIng commented:
@CedArctic 
Excellent, I'll run a final test tomorrow and then merge. I'll open a new PR to add DietPi-Services support afterwards.

CedArctic commented:
@MichaIng Sounds great! Let me know how the testing goes and if we need to enable force_wsgi_environ in the configuration file after all.

MichaIng commented:
Testing:
```
G_CONFIG_INJECT 'DEV_GITOWNER=' 'DEV_GITOWNER=CedArctic' /boot/dietpi.txt
G_CONFIG_INJECT 'DEV_GITBRANCH=' 'DEV_GITBRANCH=dev' /boot/dietpi.txt
dietpi-update -1
dietpi-software install 177
```
- [x] Stretch
- [x] Buster

🈴 Bullseye does not have python(2)-virtualenv (yet), hence does not support Firefox Sync Server for now

MichaIng committed:
v6.31

+ DietPi-Software | Firefox Sync Server: Disable on Bullseye, since python(2)-virtualenv is not available there (yet)
+ DietPi-Software | Firefox Sync Server: Fix wrong archive directory name
+ DietPi-Software | Firefox Sync Server: Add documentation link to service file
+ DietPi-Software | Firefox Sync Server: Remove ffsync group as well on uninstall
+ DietPi-Software | Firefox Sync Server: Adding cosmetic/consistent tabs and empty lines

MichaIng committed:
v6.31

+ DietPi-Software | Firefox Sync Server: ffsync user is not added to dietpi group anymore, hence remove invalid "-a" option from usermod

MichaIng commented:
@CedArctic 
Okay, fixed mainly a wrong download directory name, now works fine, Bullseye needed to be disabled for now as above.

I wanted to test with Firefox, but I don't have a Mozilla account and authentication against Mozilla account server is still required.

Found out that one can run a Mozilla authentication server as well: https://github.com/mozilla/fxa
Respectively for authentication server only: https://github.com/mozilla/fxa/tree/master/packages/fxa-auth-server
which depends on: https://github.com/mozilla/fxa/tree/master/packages/fxa-auth-db-mysql

Seems to be a larger task, however at least I already found the Firefox settings to adjust this as well:
- identity.fxaccounts.auth.uri
- identity.fxaccounts.remote.oauth.uri (not sure if OAuth is included with above FxA auth server)

Worth it do start working on this as well?
_________
Running a test on ARMv8 to cover pre-compiling...

CedArctic commented:
@MichaIng yes I was also considering taking a look into self-hosting an authentication server however I think I came across some info about difficulties when it came to running on ARM platforms. Definitely worth taking a look into all this and possibly adding the authentication server integration.

MichaIng commented:
@CedArctic 
Okay, lets make this an own issue/request. Test on ARMv8 went well, will merge.
